packages/shorebird_cli/lib/src/artifact_builder/flutter_build_process_tracker.dart

```diff
@@ -0,0 +1,136 @@
+import 'dart:collection';
+
+import 'package:mason_logger/mason_logger.dart';
+import 'package:shorebird_cli/src/logger.dart';
+import 'package:shorebird_cli/src/platform.dart';
+import 'package:shorebird_cli/src/shorebird_process.dart';
+
+/// This enum describes the different steps that the log updater
+/// identifies in the build process.
+///
+/// Not all of the steps apply to all platforms or to the situation of the
+/// user's build (e.g. [downloadingGradleW] only applies to Android projects
+/// in the first build).
+///
+/// Refer to the comment of each step for more information.
+enum FlutterBuildProcessTrackerStep {
+  /// The build has just started.
+  initial,
+
+  /// Generic step called when the build is happening, but
+  /// no specifc, or outstanding, step has been identified.
+  building,
+
+  /// The project does not have a gradle wrapper binary and it is being
+  /// downloaded. (Android only)
+  downloadingGradleW,
+
+  /// The project is downloading a Gradle dependency. (Android only)
+  downloadingGradleDependency,
+
+  /// Flutter is checking and preparing the Android SDK. (Android only)
+  preparingAndroidSDK,
+
+  /// The flutter assemble command is executing.
+  flutterAssemble;
+
+  /// A message that describes the step.
+  ///
+  /// When a step message should be defined by the user interface, this
+  /// method will return null.
+  String? get message {
+    switch (this) {
+      case FlutterBuildProcessTrackerStep.downloadingGradleW:
+        return 'Downloading Gradle Wrapper';
+      case FlutterBuildProcessTrackerStep.downloadingGradleDependency:
+        return 'Downloading Gradle Dependency';
+      case FlutterBuildProcessTrackerStep.preparingAndroidSDK:
+        return 'Preparing Android SDK';
+      case FlutterBuildProcessTrackerStep.flutterAssemble:
+        return 'Assembling Flutter Resources';
+      case FlutterBuildProcessTrackerStep.initial:
+      case FlutterBuildProcessTrackerStep.building:
+        return null;
+    }
```

Comment:
I think as we add more, you're going to want these to be closer to the detection logic and name.  I suspect you'll end up with a system like:

final steps = [
   step('Downloading Gradle Wrapper', (line) => line.contains('Downloading') && line.contains('.gradle))
...
];
Or something. 🤷